dsgrid/registry/project_registry_manager.py

```diff
@@ -239,36 +279,69 @@ def _submit_dataset(
+        mapping_references: List[DimensionMappingReferenceModel],
+    ):
+        """Check that a dataset has all project-required dimension records."""
+        handler = make_dataset_schema_handler(
+            dataset_config, self._dimension_mgr, self._dimension_mapping_mgr
+        )
+        pivot_dimension = handler.get_pivot_dimension_type()
+        exclude_dims = set([DimensionType.TIME, DimensionType.DATA_SOURCE, pivot_dimension])
+        types = [x for x in DimensionType if x not in exclude_dims]
+        dimension_pairs = [tuple(sorted((x, y))) for x, y in itertools.combinations(types, 2)]
+
+        data_source = dataset_config.model.data_source
+        associations = project_config.dimension_associations
+        dim_table = handler.get_unique_dimension_rows(mapping_references)
+        for type1, type2 in dimension_pairs:
+            records = associations.get_associations_by_data_source(data_source, type1, type2)
+            if records is None:
+                records = self._get_project_dimensions_table(project_config, type1, type2)
+            columns = (type1.value, type2.value)
+            with Timer(timer_stats_collector, "evaluate dimension record counts"):
+                record_count = records.count()
+                count = records.select(*columns).intersect(dim_table.select(*columns)).count()
+            if count != record_count:
+                table = records.select(*columns).exceptAll(dim_table.select(*columns))
+                dataset_id = dataset_config.model.dataset_id
+                with io.StringIO() as buf, redirect_stdout(buf):
+                    table.show(n=table.count())
+                    logger.error(
+                        "Dataset %s is missing records for %s:\n%s",
+                        dataset_id,
+                        (type1, type2),
+                        buf.getvalue(),
+                    )
+                raise DSGInvalidDataset(
+                    f"Dataset {dataset_id} is missing records for {(type1, type2)}"
+                )
+        self._check_pivot_dimension_columns(
+            project_config, handler, pivot_dimension, mapping_references
+        )
+
+    @staticmethod
+    def _get_project_dimensions_table(project_config, type1, type2):
+        pdim1 = project_config.get_base_dimension(type1)
+        pdim1_ids = pdim1.get_unique_ids()
+        pdim2 = project_config.get_base_dimension(type2)
+        pdim2_ids = pdim2.get_unique_ids()
+        records = itertools.product(pdim1_ids, pdim2_ids)
+        return create_dataframe_from_dimension_ids(records, type1, type2)
+
+    @staticmethod
+    def _check_pivot_dimension_columns(project_config, handler, dim_type, mapping_refs):
+        d_dim_ids = set(
+            handler.get_pivot_dimension_columns_mapped_to_project(mapping_refs).values()
+        )
+        p_dim_ids = project_config.get_base_dimension(dim_type).get_unique_ids()
+        diff = d_dim_ids.difference(p_dim_ids)
```

Comment:
We want symmetric_difference here right since we are checking that the mapped dimensions match exactly what's expected at project?

Reply:
If not symmetric_difference, I think it's `diff = p_dim_ids.difference(d_dim_ids)`

Reply:
I'm not sure how I messed this one up. The exact project-required records are known here. I'll fix.